File: packages/vscode/test/tools.test.ts
/**
 * Unit tests for the tool result payload structures.
 */

import { deepStrictEqual } from 'assert';

interface ToolResultPayload {
  ok: boolean | null;
  applied?: boolean;
  content?: string;
  error?: string;
  files?: Array<{ name: string; path: string; type: 'file' | 'directory' }>;
  path?: string;
}

// ── Tests ──────────────────────────────────────────────────────────────

// Successful file read
{
  const result: ToolResultPayload = {
    ok: true,
    content: 'export function hello() { return "world"; }',
  };
  deepStrictEqual(result.ok, true);
  deepStrictEqual(result.content?.length, 43);
  deepStrictEqual(result.error, undefined);
}

// Failed file read
{
  const result: ToolResultPayload = {
    ok: false,
    error: 'File not found: src/missing.ts',
  };
  deepStrictEqual(result.ok, false);
  deepStrictEqual(result.error, 'File not found: src/missing.ts');
}

// Successful edit
{
  const result: ToolResultPayload = {
    ok: true,
    applied: true,
  };
  deepStrictEqual(result.ok, true);
  deepStrictEqual(result.applied, true);
}

// Rejected edit
{
  const result: ToolResultPayload = {
    ok: false,
    applied: false,
    error: 'User rejected the change.',
  };
  deepStrictEqual(result.ok, false);
  deepStrictEqual(result.applied, false);
}

// Deferred tool call (ok is null — not yet executed)
{
  const result: ToolResultPayload = {
    ok: null,
    path: 'src/middleware/',
  };
  deepStrictEqual(result.ok, null);
  deepStrictEqual(result.path, 'src/middleware/');
}

// Directory browse
{
  const result: ToolResultPayload = {
    ok: true,
    files: [
      { name: 'login.ts', path: 'src/login.ts', type: 'file' },
      { name: 'middleware', path: 'src/middleware', type: 'directory' },
    ],
  };
  deepStrictEqual(result.ok, true);
  deepStrictEqual(result.files?.length, 2);
  deepStrictEqual(result.files![0]!.type, 'file');
  deepStrictEqual(result.files![1]!.type, 'directory');
}

// Command result with error
{
  const result: ToolResultPayload = {
    ok: false,
    content: 'npm ERR! missing script: build',
    error: 'Command failed with exit code 1',
  };
  deepStrictEqual(result.ok, false);
  ok(result.content!.includes('npm ERR!'));
}

function ok(condition: boolean, message?: string): void {
  deepStrictEqual(condition, true, message);
}

console.log('✅ Tool result tests passed');

File: packages/jetbrains/src/test/kotlin/com/hermes/ide/ToolsTest.kt
package com.hermes.ide

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Assertions.*

/**
 * Unit tests for tool result payload structures.
 *
 * Tests the ToolResultPayload and related data class shapes.
 * Mirrors the VS Code tools.test.ts.
 */
class ToolsTest {

    @Test
    fun `successful file read result`() {
        val result = ToolResultPayload(
            ok = true,
            content = """export function hello() { return "world"; }""",
        )
        assertEquals(true, result.ok)
        assertEquals(43, result.content?.length)
        assertNull(result.error)
    }

    @Test
    fun `failed file read result`() {
        val result = ToolResultPayload(
            ok = false,
            error = "File not found: src/missing.ts",
        )
        assertEquals(false, result.ok)
        assertEquals("File not found: src/missing.ts", result.error)
    }

    @Test
    fun `successful edit result`() {
        val result = ToolResultPayload(
            ok = true,
            applied = true,
        )
        assertEquals(true, result.ok)
        assertEquals(true, result.applied)
    }

    @Test
    fun `rejected edit result`() {
        val result = ToolResultPayload(
            ok = false,
            applied = false,
            error = "User rejected the change.",
        )
        assertEquals(false, result.ok)
        assertEquals(false, result.applied)
    }

    @Test
    fun `deferred tool call — ok is null`() {
        val result = ToolResultPayload(
            ok = null,
            path = "src/middleware/",
        )
        assertNull(result.ok)
        assertEquals("src/middleware/", result.path)
    }

    @Test
    fun `directory browse result`() {
        val result = ToolResultPayload(
            ok = true,
            files = listOf(
                FileEntry(name = "login.ts", path = "src/login.ts", type = "file"),
                FileEntry(name = "middleware", path = "src/middleware", type = "directory"),
            ),
        )
        assertEquals(true, result.ok)
        assertEquals(2, result.files?.size)
        assertEquals("file", result.files?.get(0)?.type)
        assertEquals("directory", result.files?.get(1)?.type)
    }

    @Test
    fun `command result with error`() {
        val result = ToolResultPayload(
            ok = false,
            content = "npm ERR! missing script: build",
            error = "Command failed with exit code 1",
        )
        assertEquals(false, result.ok)
        assertTrue(result.content?.contains("npm ERR!") == true)
    }
}
